### ISSUE TITLE
RFC - 6.0 - Allow routes being mapped without a controller

### ISSUE THREAD

**Issue by LordSimal** [cakephp/cakephp#18065]
### Description

As reported in slack by Erwane there is currently no way to do simple URL mapping without a Controller like the following examples:
```
$routes->get('info', function() { 
    phpinfo(); 
});
```
or
```
$routes->connect('up', function() { 
    return new Response('body' => '', 'status' => 200);
});
```

Sure, one could either re-use their AppController (or a custom Controller) and add those methods + map that action in the Router, but I think it would allow for a bit less coupling to have a simple HTTP Request => Response

Here are some things that came to my mind related to the possible behaviour of this feature:

* since there is no view object present, no layout can/should be rendered
* echo'ed content should be returned as is (e.g. `phpinfo();`) via e.g. output buffering
* if a Response object is returned, it contains all the info which should be returned. Content from output buffering should be ignored.
* this would not work with authentication/authorization (unless the RoutingMiddleware is put after those middlewares of course) but I'd argue this kind of approach won't be used in an auth context.

The [current param types](https://github.com/cakephp/cakephp/blob/5.x/src/Routing/RouteBuilder.php#L443C43-L443C63) in the Router prevents this feature from being added in a minor.

### CakePHP Version

6.0

**Comment by markstory** [comment]
I have a few questions about the proposed interface.

- Would the route functions be passed `$request` as a parameter? 
- Would` other DI services also be injected into these functions?
- How do route functions interact with middleware that is attached below `RoutingMiddleware`? Are those middleware layers skipped/not called?

**Comment by LordSimal** [comment]
> Would the route functions be passed `$request` as a parameter?
> Would` other DI services also be injected into these functions?

No, either one would be an indicator, that a Controller action should be used. This should just be a shortcut for very simple routes.

> How do route functions interact with middleware that is attached below RoutingMiddleware? Are those middleware layers skipped/not called?

I'd expect, that route specific middlewares should still work the same as they do now.
So
```
$routes->scope('/blog', function (RouteBuilder $routes) {
    $routes->applyMiddleware('auth');
    $routes->get('info', function() { 
        phpinfo(); 
    });
});
```
would only work for logged in users.

**Comment by markstory** [comment]
> I'd expect, that route specific middlewares should still work the same as they do now.
> So ... would only work for logged in users.

That makes sense to me. I was more wondering about middleware like

```php
// in Application::middleware()

            ->add(new RoutingMiddleware($this))

            // Parse various types of encoded request bodies so that they are
            // available as array through $request->getData()
            // https://book.cakephp.org/4/en/controllers/middleware.html#body-parser-middleware
            ->add(new BodyParserMiddleware())

            ->add(new CspMiddleware($this->getContainer()->get(CSPBuilder::class)))
            ->add(new AuthenticationMiddleware($this))
            ->add(new AuthorizationMiddleware($this, [
                'identityDecorator' => function ($auth, $user) {
                    return $user->setAuthorization($auth);
                },
            ]));
```

Since the responses are being handled by the `RoutingMiddleware,` no other middleware below `RoutingMiddleware` would execute as the response would have been returned.

> No, either one would be an indicator, that a Controller action should be used. This should just be a shortcut for very simple routes.

Perhaps we revisit that in the future? I can see us getting feedback as desired features. Personally, I can see this style of API being useful for building small applications, you could compose applications together via includes. You could put most of the 'domain' logic into services and have a files with collections of callables bound to URLs. It could look a bit like `flask` if we were thoughtful about it 🤔

**Comment by umer936** [comment]
Makes me think of the routes in express.js or some of the other js webserver routing. 

It could be helpful, but also it seems like there's a few ways to do more or less the same thing. For `phpinfo();`, I just put it in `webroot/index.php` if my app isn't working. Or if it is, just have it in AppController, PagesController, etc. 

I think it's a cool idea, and if the implementation isn't hard, it might be nice. However, I also can't think of any places where I would use this. 

Maybe for a sort of "API" that just returns specific PHP responses? It'd be quick in creating it as well as serving it since it doesn't call up anything else. 

I currently have a bun server that basically does this: 
```js
  if (pathname === "/about" || pathname === "/hapi/about") {
      const aboutFile = Bun.file("./HAPI-server/about.json");
      return new Response(aboutFile, {
          headers: {
              "Content-Type": "application/json",
          }
      });
  };
  if (pathname === "/capabilities" || pathname === "/hapi/capabilities") {
      const aboutFile = Bun.file("./HAPI-server/capabilities.json");
      return new Response(aboutFile, {
          headers: {
              "Content-Type": "application/json",
          }
      });
  };
  if (pathname === "/catalog" || pathname === "/hapi/catalog") {
      const aboutFile = Bun.file("./HAPI-server/catalog.json");
      return new Response(aboutFile, {
          headers: {
              "Content-Type": "application/json",
          }
      });
  };
```

Don't know if that would be the kind of thing to target or if those are "too involved" for a route response like this. Even if this proposed change were there, I probably would have used either JS or Python for my cases, but I can see why someone might want to return "static-_ish_" (i.e. non-database, etc) responses. 

I'm curious other usecase ideas.

**Comment by ADmad** [comment]
> However, I also can't think of any places where I would use this.

Same here, this might not be too difficult to implement but I would like to hear a few concrete use cases for it.
